### DISCUSSION TITLE
how to support progressPlugin?

### DISCUSSION THREAD

**Discussion by hardfist** [web-infra-dev/rspack#4202] [Core]
webpack's internal progressPlugin is not pretty and https://www.npmjs.com/package/webpackbar is much pretty.
so which behavior and interface should we align to?
we have needs for custom color but webpack's internal progressPlugin doesn't support it and webpackbar supports custom color, should we add color option to internal rspack.progressPlugin?

**Comment by hardfist** [comment]
cc @h-a-n-a @ahabhgk @chenjiahan @Hamzakh777

**Comment by chenjiahan** [comment]
Can Rspack implement the `webpack.ProgressPlugin`?

**Reply by Hamzakh777** [reply]
@chenjiahan I believe Rspack already supports `ProgressPlugin`, no? 

We can support colors on `rspack.ProgressPlugin`. We can add a prop called color. But currently, we set the color for the bar in the format `bar:.color1/color2` [here](https://github.com/web-infra-dev/rspack/blob/main/crates/rspack_plugin_progress/src/lib.rs#L33), where `color1` is used to the elapsed part, and `color2` is used for the bar that is yet to render. The colors we can support are the ones defined in the `enum Color` for the `console` crate https://docs.rs/console/latest/console/enum.Color.html.

**Reply by chenjiahan** [reply]
I mean the JS plugin, the same as https://webpack.js.org/plugins/progress-plugin/#root

Although we can add some options to the Rust progress plugin, the behaviour is still not consistent with webpack. If we can implement the JS plugin, it will be nice for webpack users.

**Reply by chenjiahan** [reply]
Related issue: https://github.com/web-infra-dev/rspack/issues/3678

**Reply by ahabhgk** [reply]
Maybe both? ProgressPlugin for the default one, ProgressRspackPlugin for the enhanced one

**Reply by hardfist** [reply]
I'm not sure whether it's worthwhile maintain two progressPlugin

**Reply by ahabhgk** [reply]
then ProgressPlugin with a "bar" options, we can't compatible with JS side ProgressPlugin since it needs too many JS <-> Rust overhead anyway

**Reply by chenjiahan** [reply]
It would be nice if Rspack could provide a simplified JS side ProgressPlugin.

There are still many differences between the Modern.js custom progress bar and the Rspack built in progress bar.

**Reply by Hamzakh777** [reply]
If possible, can you explain what do you mean by the `JS side ProgressPlugin`?

**Reply by chenjiahan** [reply]
@Hamzakh777 A JS side plugin is just like a webpack plugin, it can be consumed by Node.js users and can be registered in the `rspack.config.*` file.

**Reply by Hamzakh777** [reply]
Thanks @chenjiahan! In Rspack, how does the execution differ from an `Rspack plugin` to a `JS plugin`? I thought even the JS plugin gets compiled and executed on the Rspack side.

**Reply by chenjiahan** [reply]
> In Rspack, how does the execution differ from an Rspack plugin to a JS plugin? I thought even the JS plugin gets compiled and executed on the Rspack side.

@hardfist Does Rspack have a design document for the JS side plugin?

**Reply by Hamzakh777** [reply]
I don't think there is one, @hardfist?

**Reply by hardfist** [reply]
i will try add one today

**Comment by chenjiahan** [comment]
To provide more context, I will list the differences between the Modern.js custom progress plugin and the Rspack builtin progress plugin:

1. The custom progress can be fixed in the bottom of the terminal, this can avoid the progress bar being broken but other logs, and it is also more friendly to display multiple progress bars.

![image](https://github.com/web-infra-dev/rspack/assets/7237365/8ae94f07-4522-43ad-8c9e-8cca5bf71f3f)

2. The custom progress bar can display multiple progress bars correctly, while the Rspack progress bar still has a bug, ref: https://github.com/web-infra-dev/rspack/issues/4214.

![image](https://github.com/web-infra-dev/rspack/assets/7237365/13a0f53f-afb2-4a30-9aea-6ea8de32059d)

3. The custom progress bar can display different colors for different targets. For example, green for the client and cyan for the server.

4. The custom progress bar can be hidden after the compilation is done.

![image](https://github.com/web-infra-dev/rspack/assets/7237365/5caf5c8e-864e-4953-92e1-6ec7cef25c8a)

5. The custom progress bar can display plain text in no-TTY environment (such as CI).

![image](https://github.com/web-infra-dev/rspack/assets/7237365/cf7922d9-21e8-46dd-b02e-7599a259cb33)

6. The custom progress bar can simulate the progress rate to avoid progress regression problems, whereas the Rspack progress rate is unstable and can drop to a lower value.
